aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -152,32 +161,40 @@ void multi_tensor_apply(
     T callable,
     ArgTypes... args) {
         TORCH_CHECK(tensor_lists.size() == depth, "Number of tensor lists has to match the depth.");
-        size_t n_tensors = tensor_lists[0].size();
+        const size_t n_tensors = tensor_lists[0].size();
+        size_t n_zero_tensors = 0;
         TensorListMetadata<depth> tensorListMeta;
         tensorListMeta.start_tensor_this_launch = 0;
 
         int loc_block_info = 0;
         int loc_tensor_info = 0;
         for(size_t t = 0; t < n_tensors; t++) {
             if (tensor_lists[0][t].numel() == 0) {
-                continue;
-            }
-            tensorListMeta.numel_for_tensor[loc_tensor_info] = tensor_lists[0][t].numel();
-            for (int d = 0; d < depth; d++) {
-                tensorListMeta.addresses[d][loc_tensor_info] = tensor_lists[d][t].const_data_ptr();
+                n_zero_tensors++;
+                if (t != n_tensors - 1) {
```

Comment:
I see this is built on top of existing code, but the various "t == n_tensors - 1" equality checks (in 174, 188, and 197) are not clean. I'm thinking we should be able to break this code into a for loop iterating through tensors that update the n_zero_tensors, tensorListMeta and chunk information followed by the `multi_tensor_apply_kernel` (which is only run on the last chunk anyway, so it really should not live in the loop).

Reply:
I'm not sure if that will be completely equivalent from what I understand

Reply:
I haven't been good enough to do that. If you've got something working already, feel free to take over this

Reply:
If it's a hard refactor, we should land as is, because it fixes a major bug, sorry, I don't have anything working.

Reply:
Sounds reasonable to me, I'll file an issue to just track that this as a valuable next step. https://github.com/pytorch/pytorch/issues/101096